The function "parseCidr" was not found

Migrated from https://github.com/microsoft/PSRule/issues/1549

---

**Description of the issue**

New Bicep function `parseCidr` is not supported.

**To Reproduce**

Steps to reproduce the issue:

1. Add Bicep resource for SQL Server Firewall Rules
2. Reference function `parseCidr` to process incoming values
3. Notice error when running PSRule tests.

**Bicep module**
```bicep
@description('The name of the SQL Server.')
param SqlServerName string

@description('''
An object containing the list of public IP Addresses that are allowed to access the Resource (Name, CIDR).
**IMPORTANT:** The `CIDR` value must be an `array` of strings rather than a single comma-separated string.
''')
param AllowedPublicIp object

resource sqlServer 'Microsoft.Sql/servers@2022-05-01-preview' existing = {
  name: SqlServerName
}

resource firewallRules 'Microsoft.Sql/servers/firewallRules@2022-05-01-preview' = [for (ip, index) in AllowedPublicIp.CIDR: {
  name: '${AllowedPublicIp.Name}_${index}'
  parent: sqlServer
  properties: {
    startIpAddress: parseCidr(ip).firstUsable
    endIpAddress: parseCidr(ip).lastUsable
  }
}]
```
**Bicep Params**
```bicep
using './main.bicep'

param SqlServerName = 'some-existing-sql-server'
param AllowedPublicIp = {
  Name: 'fake_ip'
  CIDR: [ '10.1.2.0/23', '10.10.2.0/23' ]
  Priority: 100
}
```

**Expected behaviour**

The Bicep module should expand and get processed.

**Error output**

```text
[ERROR] Failed to expand bicep source 'C:\source\sql-firewallrule\main.tests.bicep'.
Exception calling "GetBicepResources" with "2" argument(s):
Unable to expand resources because the source file 'C:\source\sql-firewallrule\main.tests.bicep' was not valid.
An error occurred evaluating expression '[parseCidr(parameters('AllowedPublicIp').CIDR[copyIndex()]).firstUsable]' line 72.
The function "parseCidr" was not found.
```

**Module in use and version:**

- Module: PSRule
- Version: **2.9.0**
- Module: PSRule.Rules.Azure
- Version: **1.27.2**
- Module: Bicep
- Version: **0.18.4**

Captured output from `$PSVersionTable`:

```text
Name                           Value
----                           -----
PSVersion                      7.3.4
PSEdition                      Core
GitCommitId                    7.3.4
OS                             Microsoft Windows 10.0.19045
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

**Additional context**

The `parseCidr` function is relatively new (Bicep 0.17.1) - perhaps it's just not implemented in PSRule, or maybe there's a phantom older version of Bicep hiding on my machine?

--- a/docs/CHANGELOG-v1.md
+++ b/docs/CHANGELOG-v1.md
@@ -24,6 +24,13 @@ See [upgrade notes][1] for helpful information when upgrading from previous vers
 
 ## Unreleased
 
+What's changed since pre-release v1.28.0-B0079:
+
+- General improvements:
+  - Added support for Bicep CIDR functions by @BernieWhite.
+    [#2279](https://github.com/Azure/PSRule.Rules.Azure/issues/2279)
+    - Added support for `parseCidr`, `cidrSubnet`, and `cidrHost`.
+
 ## v1.28.0-B0079 (pre-release)
 
 What's changed since pre-release v1.28.0-B0045:

--- a/src/PSRule.Rules.Azure/Data/Template/CidrParsing.cs
+++ b/src/PSRule.Rules.Azure/Data/Template/CidrParsing.cs
@@ -1,0 +1,368 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System;
+using System.Text;
+using System.Threading;
+using Newtonsoft.Json;
+using PSRule.Rules.Azure.Resources;
+
+namespace PSRule.Rules.Azure.Data.Template
+{
+    /// <summary>
+    /// Defines information for a CIDR.
+    /// </summary>
+    /// <remarks>
+    /// See <seealso href="https://learn.microsoft.com/dotnet/fundamentals/networking/ipv6-overview"/>.
+    /// </remarks>
+    internal interface ICidrInfo
+    {
+        public string Broadcast { get; }
+        public int Cidr { get; }
+        public string FirstUsable { get; }
+        public string LastUsable { get; }
+        public string Netmask { get; }
+        public string Network { get; }
+    }
+
+    internal static class CidrParsing
+    {
+        // Parsing and formatting constants
+        private const char SLASH = '/';
+        private const char COLON = ':';
+        private const char DOT = '.';
+        private const char ZERO = '0';
+        private const string COLON_STRING = ":";
+
+        // Format strings
+        private const string FORMAT_HEX = "x";
+        private const string FORMAT_2HEX = "x2";
+
+        internal struct CidrIP
+        {
+            public CidrIP(byte[] value)
+            {
+                Value = FormatIPAddress(value);
+                Bytes = value;
+                Lo = 0;
+                Hi = 0;
+
+                // IPv4
+                if (value.Length == 4)
+                {
+                    for (var i = 0; i < 4; i++)
+                    {
+                        Lo |= (ulong)value[i] << (8 * i);
+                    }
+                }
+                // IPv6
+                else
+                {
+                    for (var i = 0; i < 8; i++)
+                    {
+                        Hi |= (ulong)value[i] << (8 * i);
+                        Lo |= (ulong)value[i + 8] << (8 * i);
+                    }
+                }
+            }
+
+            public byte[] Bytes { get; }
+
+            public string Value { get; }
+
+            public ulong Hi { get; }
+            public ulong Lo { get; }
+        }
+
+        internal sealed class CidrInfo : ICidrInfo
+        {
+            [JsonIgnore]
+            public int Version { get; internal set; }
+
+            [JsonIgnore]
+            public CidrIP Network { get; internal set; }
+
+            [JsonIgnore]
+            public CidrIP Netmask { get; internal set; }
+
+            [JsonIgnore]
+            public CidrIP Broadcast { get; internal set; }
+
+            [JsonIgnore]
+            public CidrIP FirstUsable { get; internal set; }
+
+            [JsonIgnore]
+            public CidrIP LastUsable { get; internal set; }
+
+            [JsonProperty("cidr", Order = 1)]
+            public int Cidr { get; internal set; }
+
+            [JsonProperty("broadcast", Order = 0, NullValueHandling = NullValueHandling.Ignore)]
+            string ICidrInfo.Broadcast => Version == 4 ? Broadcast.Value : null;
+
+            [JsonProperty("firstUsable", Order = 2)]
+            string ICidrInfo.FirstUsable => FirstUsable.Value;
+
+            [JsonProperty("lastUsable", Order = 3)]
+            string ICidrInfo.LastUsable => LastUsable.Value;
+
+            [JsonProperty("netmask", Order = 4)]
+            string ICidrInfo.Netmask => Netmask.Value;
+
+            [JsonProperty("network", Order = 5)]
+            string ICidrInfo.Network => Network.Value;
+        }
+
+        public static bool TryParse(string value, out ICidrInfo info)
+        {
+            var cidrParts = value.Split(SLASH);
+            var addressBytes = GetAddressBytes(cidrParts[0]);
+            var length = int.Parse(cidrParts[1]);
+            return TryGetInfo(addressBytes, length, 0, 0, out info);
+        }
+
+        public static bool TryGetSubnet(string value, int prefixLength, int subnetOffset, out string subnet)
+        {
+            if (prefixLength < 0 || prefixLength > 128)
+                throw new ExpressionArgumentException(string.Format(Thread.CurrentThread.CurrentCulture, PSRuleResources.ArgumentInvalidCIDRLength, prefixLength, value));
+
+            subnet = null;
+            var cidrParts = value.Split(SLASH);
+            var addressBytes = GetAddressBytes(cidrParts[0]);
+            if (!TryGetInfo(addressBytes, prefixLength, subnetOffset, 0, out var info))
+                return false;
+
+            if (prefixLength < info.Cidr)
+                throw new ExpressionArgumentException(string.Format(Thread.CurrentThread.CurrentCulture, PSRuleResources.ArgumentInvalidCIDRLength, prefixLength, value));
+
+            subnet = string.Concat(info.Network, SLASH, prefixLength);
+            return true;
+        }
+
+        public static bool TryGetHost(string value, int hostOffset, out string host)
+        {
+            host = null;
+            var cidrParts = value.Split(SLASH);
+            var addressBytes = GetAddressBytes(cidrParts[0]);
+            var length = int.Parse(cidrParts[1]);
+            if (!TryGetInfo(addressBytes, length, 0, addressBytes.Length == 4 ? hostOffset : hostOffset + 1, out var info))
+                return false;
+
+            host = info.FirstUsable;
+            return true;
+        }
+
+        private static bool TryGetInfo(byte[] address, int prefixLength, int subnetOffset, int hostOffset, out ICidrInfo info)
+        {
+            var numBytes = address.Length;
+            var offset = (byte)(numBytes == 4 ? 1 : 0);
+
+            // Fill in the mask
+            var netmask = new byte[numBytes];
+            for (var i = 0; i < prefixLength; i++)
+            {
+                netmask[i / 8] |= (byte)(1 << (7 - (i % 8)));
+            }
+
+            // Mask the address to get the address
+            var network = new byte[numBytes];
+            for (var i = 0; i < numBytes; i++)
+            {
+                network[i] = (byte)(address[i] & netmask[i]);
+            }
+
+            // Calculate the network based on the netmask and offset by subnetOffset
+            var networkAddress = new byte[numBytes];
+            var carry = subnetOffset << (8 - (prefixLength % 8));
+            for (var i = numBytes - (numBytes - prefixLength / 8); i >= 0; i--)
+            {
+                networkAddress[i] = (byte)((address[i] & network[i]) + carry);
+                carry = (byte)(((address[i] & network[i]) + carry) >> 8);
+            }
+
+            // Calculate the first address based on netmask
+            var firstAddress = new byte[numBytes];
+            for (var i = 0; i < numBytes; i++)
+            {
+                firstAddress[i] = (byte)(network[i] + ((i == numBytes - 1) ? (offset + hostOffset) : 0));
+            }
+
+            // Calculate the broadcast
+            var broadcastAddress = new byte[numBytes];
+            for (var i = 0; i < numBytes; i++)
+            {
+                broadcastAddress[i] = (byte)(networkAddress[i] + ((i >= prefixLength / 8) ? 255 - netmask[i] : 0));
+            }
+
+            // Calculate the last usable address based on broadcast
+            var lastUsableAddress = new byte[numBytes];
+            for (var i = 0; i < numBytes; i++)
+            {
+                lastUsableAddress[i] = (byte)(broadcastAddress[i] - ((i == numBytes - 1) ? offset : 0));
+            }
+
+            info = new CidrInfo
+            {
+                Version = numBytes == 4 ? 4 : 6,
+                Network = new CidrIP(networkAddress),
+                Netmask = new CidrIP(netmask),
+                Broadcast = new CidrIP(broadcastAddress),
+                FirstUsable = new CidrIP(firstAddress),
+                LastUsable = new CidrIP(lastUsableAddress),
+                Cidr = prefixLength,
+            };
+            return true;
+        }
+
+        static byte[] GetAddressBytes(string value)
+        {
+            if (value.Contains(COLON_STRING))
+            {
+                // IPv6
+                var ip = new byte[16];
+                var index = 0;
+                var currentByte = 0;
+                var currentByteIndex = 0;
+                var doubleColonIndex = -1;
+                for (var i = 0; i < value.Length; i++)
+                {
+                    if (value[i] == COLON)
+                    {
+                        if (i == 0 || value[i - 1] == COLON)
+                        {
+                            doubleColonIndex = index;
+                        }
+                        else
+                        {
+                            ip[currentByteIndex++] = (byte)(currentByte >> 8);
+                            ip[currentByteIndex++] = (byte)(currentByte & 0xFF);
+                            currentByte = 0;
+                        }
+                    }
+                    else
+                    {
+                        currentByte = (currentByte << 4) + ParseHexDigit(value[i]);
+                    }
+                    index++;
+                }
+                if (currentByteIndex < 16)
+                {
+                    ip[currentByteIndex++] = (byte)(currentByte >> 8);
+                    ip[currentByteIndex++] = (byte)(currentByte & 0xFF);
+                }
+                if (doubleColonIndex != -1)
+                {
+                    var numMissingBytes = 16 - currentByteIndex;
+                    for (var i = currentByteIndex - 1; i >= doubleColonIndex; i--)
+                    {
+                        ip[i + numMissingBytes] = ip[i];
+                        ip[i] = 0;
+                    }
+                }
+                return ip;
+            }
+            else
+            {
+                // IPv4
+                var ip = new byte[4];
+                var parts = value.Split(DOT);
+                for (var i = 0; i < 4; i++)
+                {
+                    ip[i] = byte.Parse(parts[i]);
+                }
+                return ip;
+            }
+        }
+
+        static int ParseHexDigit(char c)
+        {
+            if (c >= ZERO && c <= '9')
+            {
+                return c - ZERO;
+            }
+            else if (c >= 'a' && c <= 'f')
+            {
+                return c - 'a' + 10;
+            }
+            else if (c >= 'A' && c <= 'F')
+            {
+                return c - 'A' + 10;
+            }
+            throw new ArgumentException($"Invalid hex digit: {c}");
+        }
+
+        static string FormatIPAddress(byte[] ip)
+        {
+            if (ip.Length == 4)
+            {
+                // IPv4
+                return $"{ip[0]}.{ip[1]}.{ip[2]}.{ip[3]}";
+            }
+
+            // IPv6
+
+            // Find the longest sequence of consecutive zero bytes
+            var longestZeroStart = -1;
+            var longestZeroLength = 0;
+            var currentZeroStart = -1;
+            var currentZeroLength = 0;
+            for (var i = 0; i < ip.Length; i += 2)
+            {
+                if (ip[i] == 0 && ip[i + 1] == 0)
+                {
+                    if (currentZeroStart == -1)
+                    {
+                        currentZeroStart = i;
+                    }
+                    currentZeroLength += 2;
+                }
+                else
+                {
+                    if (currentZeroLength > longestZeroLength)
+                    {
+                        longestZeroStart = currentZeroStart;
+                        longestZeroLength = currentZeroLength;
+                    }
+                    currentZeroStart = -1;
+                    currentZeroLength = 0;
+                }
+            }
+            if (currentZeroLength > longestZeroLength)
+            {
+                longestZeroStart = currentZeroStart;
+                longestZeroLength = currentZeroLength;
+            }
+
+            // Build the compressed IPv6 address string
+            var sb = new StringBuilder(ip.Length);
+            for (var i = 0; i < ip.Length; i += 2)
+            {
+                if (i == longestZeroStart)
+                {
+                    sb.Append(COLON);
+                    i += longestZeroLength - 2;
+                }
+                else
+                {
+                    if (ip[i] == 0 && ip[i + 1] == 0)
+                    {
+                        sb.Append(ZERO);
+                    }
+                    else if (ip[i] == 0)
+                    {
+                        sb.Append(ip[i + 1].ToString(FORMAT_HEX));
+                    }
+                    else
+                    {
+                        sb.Append(ip[i].ToString(FORMAT_HEX));
+                        sb.Append(ip[i + 1].ToString(FORMAT_2HEX));
+                    }
+
+                    if (i < ip.Length - 2)
+                        sb.Append(COLON);
+                }
+            }
+            return sb.ToString();
+        }
+    }
+}

--- a/src/PSRule.Rules.Azure/Data/Template/ExpressionHelpers.cs
+++ b/src/PSRule.Rules.Azure/Data/Template/ExpressionHelpers.cs
@@ -11,7 +11,6 @@ using System.Security.Cryptography;
 using System.Text;
 using System.Threading;
 using Newtonsoft.Json.Linq;
-using YamlDotNet.Core.Tokens;
 
 namespace PSRule.Rules.Azure.Data.Template
 {

--- a/src/PSRule.Rules.Azure/Data/Template/Functions.cs
+++ b/src/PSRule.Rules.Azure/Data/Template/Functions.cs
@@ -163,6 +163,11 @@ namespace PSRule.Rules.Azure.Data.Template
             new FunctionDescriptor("toObject", ToObject, delayBinding: true),
             new FunctionDescriptor("lambda", Lambda, delayBinding: true),
             new FunctionDescriptor("lambdaVariables", LambdaVariables, delayBinding: true),
+
+            // CIDR
+            new FunctionDescriptor("parseCidr", ParseCidr),
+            new FunctionDescriptor("cidrSubnet", CidrSubnet),
+            new FunctionDescriptor("cidrHost", CidrHost),
         };
 
         /// <summary>
@@ -2021,6 +2026,99 @@ namespace PSRule.Rules.Azure.Data.Template
         }
 
         #endregion Lambda
+
+        #region CIDR
+
+        /// <summary>
+        /// parseCidr(network)
+        /// </summary>
+        /// <remarks>
+        /// See <seealso href="https://learn.microsoft.com/azure/azure-resource-manager/bicep/bicep-functions-cidr#parsecidr"/>.
+        /// </remarks>
+        internal static object ParseCidr(ITemplateContext context, object[] args)
+        {
+            if (CountArgs(args) != 1)
+                throw ArgumentsOutOfRange(nameof(ParseCidr), args);
+
+            if (!ExpressionHelpers.TryString(args[0], out var network))
+                throw ArgumentInvalidString(nameof(ParseCidr), "network");
+
+            try
+            {
+                if (CidrParsing.TryParse(network, out var info))
+                    return JObject.FromObject(info);
+            }
+            catch
+            {
+
+            }
+            throw new ExpressionArgumentException(string.Format(Thread.CurrentThread.CurrentCulture, PSRuleResources.ArgumentInvalidCIDR, network));
+        }
+
+        /// <summary>
+        /// cidrSubnet(network, newCIDR, subnetIndex)
+        /// </summary>
+        /// <remarks>
+        /// Splits the specified IP address range in CIDR notation into subnets with a new CIDR value and returns the IP address range of the subnet with the specified index.
+        /// See <seealso href="https://learn.microsoft.com/azure/azure-resource-manager/bicep/bicep-functions-cidr#cidrsubnet"/>.
+        /// </remarks>
+        internal static object CidrSubnet(ITemplateContext context, object[] args)
+        {
+            if (CountArgs(args) != 3)
+                throw ArgumentsOutOfRange(nameof(CidrSubnet), args);
+
+            if (!ExpressionHelpers.TryString(args[0], out var network))
+                throw ArgumentInvalidString(nameof(CidrSubnet), "network");
+
+            if (!ExpressionHelpers.TryInt(args[1], out var newCIDR))
+                throw ArgumentInvalidInteger(nameof(CidrSubnet), "newCIDR");
+
+            if (!ExpressionHelpers.TryInt(args[2], out var subnetIndex))
+                throw ArgumentInvalidInteger(nameof(CidrSubnet), "subnetIndex");
+
+            try
+            {
+                if (CidrParsing.TryGetSubnet(network, newCIDR, subnetIndex, out var subnet))
+                    return subnet;
+            }
+            catch
+            {
+
+            }
+            throw new ExpressionArgumentException(string.Format(Thread.CurrentThread.CurrentCulture, PSRuleResources.ArgumentInvalidCIDR, network));
+        }
+
+        /// <summary>
+        /// cidrHost(network, hostIndex)
+        /// </summary>
+        /// <remarks>
+        /// Calculates the usable IP address of the host with the specified index on the specified IP address range in CIDR notation.
+        /// See <seealso href="https://learn.microsoft.com/azure/azure-resource-manager/bicep/bicep-functions-cidr#cidrhost"/>.
+        /// </remarks>
+        internal static object CidrHost(ITemplateContext context, object[] args)
+        {
+            if (CountArgs(args) != 2)
+                throw ArgumentsOutOfRange(nameof(CidrHost), args);
+
+            if (!ExpressionHelpers.TryString(args[0], out var network))
+                throw ArgumentInvalidString(nameof(CidrHost), "network");
+
+            if (!ExpressionHelpers.TryInt(args[1], out var hostIndex))
+                throw ArgumentInvalidInteger(nameof(CidrHost), "hostIndex");
+
+            try
+            {
+                if (CidrParsing.TryGetHost(network, hostIndex, out var host))
+                    return host;
+            }
+            catch
+            {
+
+            }
+            throw new ExpressionArgumentException(string.Format(Thread.CurrentThread.CurrentCulture, PSRuleResources.ArgumentInvalidCIDR, network));
+        }
+
+        #endregion CIDR
 
         #region Helper functions
 

--- a/src/PSRule.Rules.Azure/Resources/PSRuleResources.Designer.cs
+++ b/src/PSRule.Rules.Azure/Resources/PSRuleResources.Designer.cs
@@ -79,6 +79,24 @@ namespace PSRule.Rules.Azure.Resources {
         }
         
         /// <summary>
+        ///   Looks up a localized string similar to The specified CIDR &apos;{0}&apos; is not valid..
+        /// </summary>
+        internal static string ArgumentInvalidCIDR {
+            get {
+                return ResourceManager.GetString("ArgumentInvalidCIDR", resourceCulture);
+            }
+        }
+        
+        /// <summary>
+        ///   Looks up a localized string similar to The specified prefix length &apos;{0}&apos; is out of range for the CIDR &apos;{1}&apos;..
+        /// </summary>
+        internal static string ArgumentInvalidCIDRLength {
+            get {
+                return ResourceManager.GetString("ArgumentInvalidCIDRLength", resourceCulture);
+            }
+        }
+        
+        /// <summary>
         ///   Looks up a localized string similar to The argument &apos;{0}&apos; for &apos;{1}&apos; is not a valid time..
         /// </summary>
         internal static string ArgumentInvalidDateTime {

--- a/src/PSRule.Rules.Azure/Resources/PSRuleResources.resx
+++ b/src/PSRule.Rules.Azure/Resources/PSRuleResources.resx
@@ -123,6 +123,12 @@
   <data name="ArgumentInvalidBoolean" xml:space="preserve">
     <value>The argument '{0}' for '{1}' is not a valid boolean.</value>
   </data>
+  <data name="ArgumentInvalidCIDR" xml:space="preserve">
+    <value>The specified CIDR '{0}' is not valid.</value>
+  </data>
+  <data name="ArgumentInvalidCIDRLength" xml:space="preserve">
+    <value>The specified prefix length '{0}' is out of range for the CIDR '{1}'.</value>
+  </data>
   <data name="ArgumentInvalidDateTime" xml:space="preserve">
     <value>The argument '{0}' for '{1}' is not a valid time.</value>
   </data>

--- a/tests/PSRule.Rules.Azure.Tests/CidrParsingTests.cs
+++ b/tests/PSRule.Rules.Azure.Tests/CidrParsingTests.cs
@@ -1,0 +1,67 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using PSRule.Rules.Azure.Data.Template;
+using Xunit;
+
+namespace PSRule.Rules.Azure
+{
+    public sealed class CidrParsingTests
+    {
+        [Fact]
+        public void IPv4()
+        {
+            Assert.True(CidrParsing.TryParse("10.0.0.0/8", out var actual));
+            Assert.Equal(8, actual.Cidr);
+            Assert.Equal("10.0.0.0", actual.Network);
+            Assert.Equal("10.0.0.1", actual.FirstUsable);
+            Assert.Equal("10.255.255.254", actual.LastUsable);
+            Assert.Equal("10.255.255.255", actual.Broadcast);
+            Assert.Equal("255.0.0.0", actual.Netmask);
+
+            Assert.True(CidrParsing.TryParse("192.168.0.0/16", out actual));
+            Assert.Equal(16, actual.Cidr);
+            Assert.Equal("192.168.0.0", actual.Network);
+            Assert.Equal("192.168.0.1", actual.FirstUsable);
+            Assert.Equal("192.168.255.254", actual.LastUsable);
+            Assert.Equal("192.168.255.255", actual.Broadcast);
+            Assert.Equal("255.255.0.0", actual.Netmask);
+
+            Assert.True(CidrParsing.TryParse("10.144.0.0/20", out actual));
+            Assert.Equal(20, actual.Cidr);
+            Assert.Equal("10.144.0.0", actual.Network);
+            Assert.Equal("10.144.0.1", actual.FirstUsable);
+            Assert.Equal("10.144.15.255", actual.Broadcast);
+            Assert.Equal("10.144.15.254", actual.LastUsable);
+            Assert.Equal("255.255.240.0", actual.Netmask);
+        }
+
+        [Fact]
+        public void IPv6()
+        {
+            Assert.True(CidrParsing.TryParse("fdad:3236:5555::/48", out var actual));
+            Assert.Equal(48, actual.Cidr);
+            Assert.Equal("fdad:3236:5555::", actual.Network);
+            Assert.Equal("fdad:3236:5555::", actual.FirstUsable);
+            Assert.Equal("fdad:3236:5555:ffff:ffff:ffff:ffff:ffff", actual.LastUsable);
+            Assert.Null(actual.Broadcast);
+            Assert.Equal("ffff:ffff:ffff::", actual.Netmask);
+
+            Assert.True(CidrParsing.TryParse("3FFE:FFFF:0:CD30:0:0:0:0/64", out actual));
+            Assert.Equal(64, actual.Cidr);
+            Assert.Equal("3ffe:ffff:0:cd30::", actual.Network);
+            Assert.Equal("3ffe:ffff:0:cd30::", actual.FirstUsable);
+            Assert.Equal("3ffe:ffff::cd30:ffff:ffff:ffff:ffff", actual.LastUsable);
+            Assert.Null(actual.Broadcast);
+            Assert.Equal("ffff:ffff:ffff:ffff::", actual.Netmask);
+
+            Assert.True(CidrParsing.TryParse("3FFE:FFFF:0:CD30::/64", out actual));
+            Assert.Equal(64, actual.Cidr);
+            Assert.Equal("3ffe:ffff:0:cd30::", actual.Network);
+            Assert.Equal("3ffe:ffff:0:cd30::", actual.FirstUsable);
+            Assert.Equal("3ffe:ffff::cd30:ffff:ffff:ffff:ffff", actual.LastUsable);
+            Assert.Null(actual.Broadcast);
+            Assert.Equal("ffff:ffff:ffff:ffff::", actual.Netmask);
+        }
+    }
+}

--- a/tests/PSRule.Rules.Azure.Tests/FunctionTests.cs
+++ b/tests/PSRule.Rules.Azure.Tests/FunctionTests.cs
@@ -30,6 +30,7 @@ namespace PSRule.Rules.Azure
         private const string TRAIT_RESOURCE = "Resource";
         private const string TRAIT_SCOPE = "Scope";
         private const string TRAIT_LAMBDA = "Lambda";
+        private const string TRAIT_CIDR = "CIDR";
 
         #region Array and object
 
@@ -1954,6 +1955,101 @@ namespace PSRule.Rules.Azure
         }
 
         #endregion Lambda
+
+        #region CIDR
+
+        [Fact]
+        [Trait(TRAIT, TRAIT_CIDR)]
+        public void ParseCidr()
+        {
+            var context = GetContext();
+
+            var actual = Functions.ParseCidr(context, new object[] { "10.144.0.0/20" }) as JObject;
+            Assert.NotNull(actual);
+            Assert.Equal("10.144.0.0", actual["network"].Value<string>());
+            Assert.Equal("255.255.240.0", actual["netmask"].Value<string>());
+            Assert.Equal("10.144.15.255", actual["broadcast"].Value<string>());
+            Assert.Equal("10.144.0.1", actual["firstUsable"].Value<string>());
+            Assert.Equal("10.144.15.254", actual["lastUsable"].Value<string>());
+            Assert.Equal(20, actual["cidr"].Value<int>());
+
+            actual = Functions.ParseCidr(context, new object[] { "fdad:3236:5555::/48" }) as JObject;
+            Assert.NotNull(actual);
+            Assert.Equal("fdad:3236:5555::", actual["network"].Value<string>());
+            Assert.Equal("ffff:ffff:ffff::", actual["netmask"].Value<string>());
+            Assert.False(actual.ContainsKeyInsensitive("broadcast"));
+            Assert.Equal("fdad:3236:5555::", actual["firstUsable"].Value<string>());
+            Assert.Equal("fdad:3236:5555:ffff:ffff:ffff:ffff:ffff", actual["lastUsable"].Value<string>());
+            Assert.Equal(48, actual["cidr"].Value<int>());
+
+            Assert.Throws<ExpressionArgumentException>(() => Functions.ParseCidr(context, null));
+            Assert.Throws<ExpressionArgumentException>(() => Functions.ParseCidr(context, new object[] { 5 }));
+            Assert.Throws<ExpressionArgumentException>(() => Functions.ParseCidr(context, new object[] { "one", "two" }));
+            Assert.Throws<ExpressionArgumentException>(() => Functions.ParseCidr(context, new object[] { "1000.144.0.0/20" }));
+            Assert.Throws<ExpressionArgumentException>(() => Functions.ParseCidr(context, new object[] { "1000.144.0.0.0.0.0.0.0.0.0.0.0.0/20" }));
+            Assert.Throws<ExpressionArgumentException>(() => Functions.ParseCidr(context, new object[] { "fdad:nnnn:5555::/48" }));
+        }
+
+        [Fact]
+        [Trait(TRAIT, TRAIT_CIDR)]
+        public void CidrSubnet()
+        {
+            var context = GetContext();
+
+            // IPv4
+            Assert.Equal("10.144.0.0/20", Functions.CidrSubnet(context, new object[] { "10.144.0.0/20", 20, 0 }) as string);
+            Assert.Equal("10.144.16.0/20", Functions.CidrSubnet(context, new object[] { "10.144.0.0/20", 20, 1 }) as string);
+            Assert.Equal("10.144.0.0/27", Functions.CidrSubnet(context, new object[] { "10.144.0.0/27", 27, 0 }) as string);
+            Assert.Equal("10.144.0.32/27", Functions.CidrSubnet(context, new object[] { "10.144.0.0/27", 27, 1 }) as string);
+            Assert.Equal("10.144.0.64/27", Functions.CidrSubnet(context, new object[] { "10.144.0.0/27", 27, 2 }) as string);
+            Assert.Equal("10.144.0.96/27", Functions.CidrSubnet(context, new object[] { "10.144.0.0/27", 27, 3 }) as string);
+            Assert.Equal("10.144.0.128/27", Functions.CidrSubnet(context, new object[] { "10.144.0.0/27", 27, 4 }) as string);
+            Assert.Equal("10.144.0.160/27", Functions.CidrSubnet(context, new object[] { "10.144.0.0/27", 27, 5 }) as string);
+            Assert.Equal("10.144.0.192/27", Functions.CidrSubnet(context, new object[] { "10.144.0.0/27", 27, 6 }) as string);
+
+            Assert.Equal("10.144.0.0/24", Functions.CidrSubnet(context, new object[] { "10.144.0.0/20", 24, 0 }) as string);
+            Assert.Equal("10.144.1.0/24", Functions.CidrSubnet(context, new object[] { "10.144.0.0/20", 24, 1 }) as string);
+            Assert.Equal("10.144.2.0/24", Functions.CidrSubnet(context, new object[] { "10.144.0.0/20", 24, 2 }) as string);
+            Assert.Equal("10.144.3.0/24", Functions.CidrSubnet(context, new object[] { "10.144.0.0/20", 24, 3 }) as string);
+            Assert.Equal("10.144.4.0/24", Functions.CidrSubnet(context, new object[] { "10.144.0.0/20", 24, 4 }) as string);
+
+            // IPv6
+            Assert.Equal("fdad:3236:5555::/52", Functions.CidrSubnet(context, new object[] { "fdad:3236:5555::/48", 52, 0 }) as string);
+            Assert.Equal("fdad:3236:5555:1000::/52", Functions.CidrSubnet(context, new object[] { "fdad:3236:5555::/48", 52, 1 }) as string);
+            Assert.Equal("fdad:3236:5555:2000::/52", Functions.CidrSubnet(context, new object[] { "fdad:3236:5555::/48", 52, 2 }) as string);
+            Assert.Equal("fdad:3236:5555:3000::/52", Functions.CidrSubnet(context, new object[] { "fdad:3236:5555::/48", 52, 3 }) as string);
+            Assert.Equal("fdad:3236:5555:4000::/52", Functions.CidrSubnet(context, new object[] { "fdad:3236:5555::/48", 52, 4 }) as string);
+
+            Assert.Throws<ExpressionArgumentException>(() => Functions.CidrSubnet(context, null));
+            Assert.Throws<ExpressionArgumentException>(() => Functions.CidrSubnet(context, new object[] { 5 }));
+            Assert.Throws<ExpressionArgumentException>(() => Functions.CidrSubnet(context, new object[] { "fdad:3236:5555::/48", 200, 0 }));
+        }
+
+        [Fact]
+        [Trait(TRAIT, TRAIT_CIDR)]
+        public void CidrHost()
+        {
+            var context = GetContext();
+
+            // IPv4
+            Assert.Equal("10.144.3.1", Functions.CidrHost(context, new object[] { "10.144.3.0/24", 0 }) as string);
+            Assert.Equal("10.144.3.2", Functions.CidrHost(context, new object[] { "10.144.3.0/24", 1 }) as string);
+            Assert.Equal("10.144.3.3", Functions.CidrHost(context, new object[] { "10.144.3.0/24", 2 }) as string);
+            Assert.Equal("10.144.3.4", Functions.CidrHost(context, new object[] { "10.144.3.0/24", 3 }) as string);
+            Assert.Equal("10.144.3.5", Functions.CidrHost(context, new object[] { "10.144.3.0/24", 4 }) as string);
+
+            // IPv6
+            Assert.Equal("fdad:3236:5555:3000::1", Functions.CidrHost(context, new object[] { "fdad:3236:5555:3000::/52", 0 }) as string);
+            Assert.Equal("fdad:3236:5555:3000::2", Functions.CidrHost(context, new object[] { "fdad:3236:5555:3000::/52", 1 }) as string);
+            Assert.Equal("fdad:3236:5555:3000::3", Functions.CidrHost(context, new object[] { "fdad:3236:5555:3000::/52", 2 }) as string);
+            Assert.Equal("fdad:3236:5555:3000::4", Functions.CidrHost(context, new object[] { "fdad:3236:5555:3000::/52", 3 }) as string);
+            Assert.Equal("fdad:3236:5555:3000::5", Functions.CidrHost(context, new object[] { "fdad:3236:5555:3000::/52", 4 }) as string);
+
+            Assert.Throws<ExpressionArgumentException>(() => Functions.CidrHost(context, null));
+            Assert.Throws<ExpressionArgumentException>(() => Functions.CidrHost(context, new object[] { 5 }));
+        }
+
+        #endregion CIDR
 
         #region Complex scenarios
 

--- a/tests/PSRule.Rules.Azure.Tests/Tests.Bicep.13.child.bicep
+++ b/tests/PSRule.Rules.Azure.Tests/Tests.Bicep.13.child.bicep
@@ -1,18 +1,20 @@
 // Copyright (c) Microsoft Corporation.
 // Licensed under the MIT License.
 
+var base = '10.0.0.0/8'
+
 var subnets = [
   {
     name: 'subnet1'
-    addressPrefix: '10.0.0.0/24'
+    addressPrefix: cidrSubnet(base, 24, 0)
   }
   {
     name: 'subnet2'
-    addressPrefix: '10.0.1.0/24'
+    addressPrefix: cidrSubnet(base, 24, 1)
   }
   {
     name: 'subnet3'
-    addressPrefix: '10.0.2.0/24'
+    addressPrefix: cidrSubnet(base, 24, 2)
   }
 ]
 

--- a/tests/PSRule.Rules.Azure.Tests/Tests.Bicep.13.json
+++ b/tests/PSRule.Rules.Azure.Tests/Tests.Bicep.13.json
@@ -1,13 +1,11 @@
 {
   "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentTemplate.json#",
-  "languageVersion": "1.10-experimental",
   "contentVersion": "1.0.0.0",
   "metadata": {
-    "_EXPERIMENTAL_WARNING": "Symbolic name support in ARM is experimental, and should be enabled for testing purposes only. Do not enable this setting for any production usage, or you may be unexpectedly broken at any time!",
     "_generator": {
       "name": "bicep",
-      "version": "0.16.2.56959",
-      "templateHash": "11969872331881049732"
+      "version": "0.18.4.5664",
+      "templateHash": "11736978922554133719"
     }
   },
   "parameters": {
@@ -75,8 +73,8 @@
     ],
     "ages": "[map(variables('dogs'), lambda('dog', lambdaVariables('dog').age))]"
   },
-  "resources": {
-    "vnet": {
+  "resources": [
+    {
       "type": "Microsoft.Resources/deployments",
       "apiVersion": "2022-09-01",
       "name": "vnet",
@@ -87,34 +85,33 @@
         "mode": "Incremental",
         "template": {
           "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentTemplate.json#",
-          "languageVersion": "1.10-experimental",
           "contentVersion": "1.0.0.0",
           "metadata": {
-            "_EXPERIMENTAL_WARNING": "Symbolic name support in ARM is experimental, and should be enabled for testing purposes only. Do not enable this setting for any production usage, or you may be unexpectedly broken at any time!",
             "_generator": {
               "name": "bicep",
-              "version": "0.16.2.56959",
-              "templateHash": "6732145744862004887"
+              "version": "0.18.4.5664",
+              "templateHash": "13364432189836274534"
             }
           },
           "variables": {
+            "base": "10.0.0.0/8",
             "subnets": [
               {
                 "name": "subnet1",
-                "addressPrefix": "10.0.0.0/24"
+                "addressPrefix": "[cidrSubnet(variables('base'), 24, 0)]"
               },
               {
                 "name": "subnet2",
-                "addressPrefix": "10.0.1.0/24"
+                "addressPrefix": "[cidrSubnet(variables('base'), 24, 1)]"
               },
               {
                 "name": "subnet3",
-                "addressPrefix": "10.0.2.0/24"
+                "addressPrefix": "[cidrSubnet(variables('base'), 24, 2)]"
               }
             ]
           },
-          "resources": {
-            "vnet": {
+          "resources": [
+            {
               "type": "Microsoft.Network/virtualNetworks",
               "apiVersion": "2022-09-01",
               "name": "vnet1",
@@ -133,17 +130,17 @@
                 ]
               }
             }
-          },
+          ],
           "outputs": {
             "subnets": {
               "type": "array",
-              "value": "[reference('vnet').subnets]"
+              "value": "[reference(resourceId('Microsoft.Network/virtualNetworks', 'vnet1'), '2022-09-01').subnets]"
             }
           }
         }
       }
     },
-    "child2": {
+    {
       "type": "Microsoft.Resources/deployments",
       "apiVersion": "2022-09-01",
       "name": "child2",
@@ -154,7 +151,7 @@
         "mode": "Incremental",
         "parameters": {
           "subnetId": {
-            "value": "[filter(reference('vnet').outputs.subnets.value, lambda('s', equals(lambdaVariables('s').name, 'subnet1')))[0].id]"
+            "value": "[filter(reference(resourceId('Microsoft.Resources/deployments', 'vnet'), '2022-09-01').outputs.subnets.value, lambda('s', equals(lambdaVariables('s').name, 'subnet1')))[0].id]"
           }
         },
         "template": {
@@ -165,8 +162,8 @@
             "_EXPERIMENTAL_WARNING": "Symbolic name support in ARM is experimental, and should be enabled for testing purposes only. Do not enable this setting for any production usage, or you may be unexpectedly broken at any time!",
             "_generator": {
               "name": "bicep",
-              "version": "0.16.2.56959",
-              "templateHash": "9874962359524344048"
+              "version": "0.18.4.5664",
+              "templateHash": "13077654233608092220"
             }
           },
           "parameters": {
@@ -178,10 +175,10 @@
         }
       },
       "dependsOn": [
-        "vnet"
+        "[resourceId('Microsoft.Resources/deployments', 'vnet')]"
       ]
     }
-  },
+  ],
   "outputs": {
     "arrayOutput": {
       "type": "array",
@@ -193,7 +190,7 @@
     },
     "vnetId": {
       "type": "string",
-      "value": "[filter(reference('vnet').outputs.subnets.value, lambda('s', equals(lambdaVariables('s').name, parameters('environment'))))[0].id]"
+      "value": "[filter(reference(resourceId('Microsoft.Resources/deployments', 'vnet'), '2022-09-01').outputs.subnets.value, lambda('s', equals(lambdaVariables('s').name, parameters('environment'))))[0].id]"
     },
     "dogNames": {
       "type": "array",

